File: plugins/expo-autofill-plugin/android-template/java/autofill/utils/RateLimitManager.java
package com.pears.pass.autofill.utils;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * Local rate limiter for the autofill master-password screen.
 * Mirrors the exponential backoff in pearpass-lib-vault-core/worklet/rateLimiter.js,
 * stored locally because the autofill plugin opens the vault store read-only.
 */
public class RateLimitManager {
    private static final String PREFS = "pp_autofill_rate_limit";
    private static final String KEY_CONSECUTIVE_FAILURES = "consecutiveFailures";
    private static final String KEY_LOCKOUT_UNTIL = "lockoutUntil";
    private static final String KEY_LAST_ATTEMPT_TIME = "lastAttemptTime";

    private static final int MAX_ATTEMPTS = 5;
    private static final long MAX_BACKOFF_MS = 24L * 60L * 60L * 1000L; // 24h
    private static final long COOLDOWN_MS = 24L * 60L * 60L * 1000L;    // 24h fresh start

    public static class Status {
        public final boolean isLocked;
        public final long lockoutRemainingMs;
        public final int remainingAttempts;

        public Status(boolean isLocked, long lockoutRemainingMs, int remainingAttempts) {
            this.isLocked = isLocked;
            this.lockoutRemainingMs = lockoutRemainingMs;
            this.remainingAttempts = remainingAttempts;
        }
    }

    private final SharedPreferences prefs;

    public RateLimitManager(Context context) {
        this.prefs = context.getApplicationContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    public Status getStatus() {
        int failures = prefs.getInt(KEY_CONSECUTIVE_FAILURES, 0);
        long lockoutUntil = prefs.getLong(KEY_LOCKOUT_UNTIL, 0L);
        long lastAttempt = prefs.getLong(KEY_LAST_ATTEMPT_TIME, 0L);
        long now = System.currentTimeMillis();

        if (lastAttempt > 0 && now - lastAttempt >= COOLDOWN_MS) {
            reset();
            return new Status(false, 0L, MAX_ATTEMPTS);
        }

        if (lockoutUntil > 0 && now >= lockoutUntil) {
            return new Status(false, 0L, 0);
        }

        boolean isLocked = lockoutUntil > 0;
        long remainingMs = isLocked ? Math.max(0L, lockoutUntil - now) : 0L;
        int remaining = isLocked ? 0 : Math.max(0, MAX_ATTEMPTS - failures);
        return new Status(isLocked, remainingMs, remaining);
    }

    public void recordFailure() {
        long now = System.currentTimeMillis();
        int failures = prefs.getInt(KEY_CONSECUTIVE_FAILURES, 0);
        long lastAttempt = prefs.getLong(KEY_LAST_ATTEMPT_TIME, 0L);
        long lockoutUntil = prefs.getLong(KEY_LOCKOUT_UNTIL, 0L);

        if (lastAttempt > 0 && now - lastAttempt >= COOLDOWN_MS) {
            failures = 0;
            lockoutUntil = 0L;
        }
        if (lockoutUntil > 0 && now >= lockoutUntil) {
            lockoutUntil = 0L;
        }

        failures++;
        long newLockoutUntil = 0L;
        if (failures >= MAX_ATTEMPTS) {
            int exponent = failures - MAX_ATTEMPTS + 1;
            long backoffMs = (long) (Math.pow(2, exponent) * 60L * 1000L);
            newLockoutUntil = now + Math.min(backoffMs, MAX_BACKOFF_MS);
        }

        prefs.edit()
                .putInt(KEY_CONSECUTIVE_FAILURES, failures)
                .putLong(KEY_LOCKOUT_UNTIL, newLockoutUntil)
                .putLong(KEY_LAST_ATTEMPT_TIME, now)
                .apply();
    }

    public void reset() {
        prefs.edit().clear().apply();
    }
}

File: plugins/expo-autofill-plugin/ios-template/PearPassAutofillExtension/RateLimitManager.swift
//
//  RateLimitManager.swift
//  PearPassAutoFillExtension
//
//  Local rate limiter for the autofill master-password screen.
//  Mirrors pearpass-lib-vault-core/worklet/rateLimiter.js. Stored locally
//  because the autofill extension opens the vault store read-only.
//

import Foundation

final class RateLimitManager {
    struct Status {
        let isLocked: Bool
        let lockoutRemainingMs: Int64
        let remainingAttempts: Int
    }

    private static let suiteName = "pp.autofill.rateLimit"
    private static let keyFailures = "consecutiveFailures"
    private static let keyLockoutUntil = "lockoutUntil"
    private static let keyLastAttempt = "lastAttemptTime"

    private static let maxAttempts = 5
    private static let maxBackoffMs: Int64 = 24 * 60 * 60 * 1000      // 24h
    private static let cooldownMs: Int64 = 24 * 60 * 60 * 1000         // 24h fresh start

    private let defaults: UserDefaults

    init() {
        self.defaults = UserDefaults(suiteName: Self.suiteName) ?? .standard
    }

    func getStatus() -> Status {
        let failures = defaults.integer(forKey: Self.keyFailures)
        let lockoutUntil = (defaults.object(forKey: Self.keyLockoutUntil) as? NSNumber)?.int64Value ?? 0
        let lastAttempt = (defaults.object(forKey: Self.keyLastAttempt) as? NSNumber)?.int64Value ?? 0
        let now = Self.nowMs()

        if lastAttempt > 0 && now - lastAttempt >= Self.cooldownMs {
            reset()
            return Status(isLocked: false, lockoutRemainingMs: 0, remainingAttempts: Self.maxAttempts)
        }

        if lockoutUntil > 0 && now >= lockoutUntil {
            return Status(isLocked: false, lockoutRemainingMs: 0, remainingAttempts: 0)
        }

        let isLocked = lockoutUntil > 0
        let remainingMs = isLocked ? max(0, lockoutUntil - now) : 0
        let remaining = isLocked ? 0 : max(0, Self.maxAttempts - failures)
        return Status(isLocked: isLocked, lockoutRemainingMs: remainingMs, remainingAttempts: remaining)
    }

    func recordFailure() {
        let now = Self.nowMs()
        var failures = defaults.integer(forKey: Self.keyFailures)
        let lastAttempt = (defaults.object(forKey: Self.keyLastAttempt) as? NSNumber)?.int64Value ?? 0
        var lockoutUntil = (defaults.object(forKey: Self.keyLockoutUntil) as? NSNumber)?.int64Value ?? 0

        if lastAttempt > 0 && now - lastAttempt >= Self.cooldownMs {
            failures = 0
            lockoutUntil = 0
        }
        if lockoutUntil > 0 && now >= lockoutUntil {
            lockoutUntil = 0
        }

        failures += 1
        var newLockoutUntil: Int64 = 0
        if failures >= Self.maxAttempts {
            let exponent = failures - Self.maxAttempts + 1
            let backoffMs = Int64(pow(2.0, Double(exponent)) * 60.0 * 1000.0)
            newLockoutUntil = now + min(backoffMs, Self.maxBackoffMs)
        }

        defaults.set(failures, forKey: Self.keyFailures)
        defaults.set(NSNumber(value: newLockoutUntil), forKey: Self.keyLockoutUntil)
        defaults.set(NSNumber(value: now), forKey: Self.keyLastAttempt)
    }

    func reset() {
        defaults.removeObject(forKey: Self.keyFailures)
        defaults.removeObject(forKey: Self.keyLockoutUntil)
        defaults.removeObject(forKey: Self.keyLastAttempt)
    }

    private static func nowMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}
